stdenv: aarch64-linux: gcc9 -> gcc12

wegank opened the pull request:
###### Description of changes

A roadblock for bumping the aarch64-linux GCC version is https://github.com/NixOS/nixpkgs/pull/167726#issuecomment-1288152384, or simply #36947. A proper fix would then be ~to add an extra stage5 to build glibc against the newer GCC.~

But, what if the bootstrap version of `libgcc_s.so` is recent enough? This PR therefore does the following:

1. Bump the GCC version to 12.
   Observe that `icu` does not build at this time.
2. `nix build -f ./pkgs/stdenv/linux/make-bootstrap-tools.nix bootstrapFiles`
   The bootstrap files build successfully.
3. Upload and use the new bootstrap files.
   Observe that `icu`, `dejagnu` and `protobuf` now build successfully.

```shell
$ nix hash file /nix/store/5vj209irqzvmzajsnj9wam1mcnrsk7ml-bootstrap-tools.tar.xz
sha256-aJvtsWeuQHbb14BGZ2EiOKzjQn46h3x3duuPEawG0eE=
$ nix hash file /nix/store/w39cbankzyc8fs5mdlikwndi3a3kgzaz-busybox
sha256-jtAPxLmSw6CoU3/CPZzZg4Q8h61EDfbz5yZXJV1cOm4=
```

Closes #108305.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

tpwrules commented:
I think it is no secret that updating the bootstraps to a more recent GCC will (probably, I haven't independently confirmed yet) work, the question is if we want to do it? I am not sure what the official process is for updating the bootstraps, or what constraints there might be on GCC versions in them.

And even right now there is a [draft PR](https://github.com/NixOS/nixpkgs/pull/206907) to upgrade to GCC 12 by default which might cause the bootstrap to need to be re-updated if there is a big change which needs a new bootstrap `libgcc_s.so` again. Unfortunately this is not really a correct solution which is why I think there has been hesitance to do it.

tpwrules commented:
Looks like this does in fact fix `icu` and avoid the compilation failure due to the mismatched `libgcc_s.so`. I will try to build my simple GUI system completely and see if anything else shows problems. Thanks for actually doing the legwork to test whether this is a viable approach.

I propose we wait and see how the GCC 12 upgrade moves. If that goes smoothly we can simply bump the bootstrap again and solve this problem for at least a little while longer. At some point pragmatism has to outweigh purity and I have heard there are problems with modern packages (particularly KDE) and GCC 9, so this really needs to be solved soon.

While that is happening, we can explore the possibility of revamping the bootstrap sequence to put this issue to bed. I don't personally feel too confident about having time/skill to do that, and there are some concerns about excessive iteration time for people who need to touch the stdenv. But it's the right thing to do in the end.

tpwrules commented:
Also worth noting, this is the last time the aarch64-linux bootstrap tarballs were updated and there is some info on people and process if we decide to go that route: https://github.com/NixOS/nixpkgs/pull/151399

K900 commented:
I think we should do this if it get us GCC 12, but I _also_ think we should really fix the bootstrap properly _before_ things explode again.

tpwrules commented:
I can confirm that a basic desktop system with Xfce and Firefox builds, boots, and seems to work okay.

One comment is that I would prefer this be one commit instead of two, with the message clarifying that the bootstrap tools in the commit are generated after the commit's other changes. This would avoid a unique and broken environment during e.g. bisection.

wegank commented:
@ofborg build stdenv stdenv.passthru.tests

gbpdt reviewed (commented):
I've tested this change and it appears to work well for me (e.g. it fixes the `icu` build problem). I'd support merging it ASAP.

tpwrules reviewed (changes requested):
Just reminding that this cannot be merged until the files are uploaded to nixos.org and the appropriate URLs inserted.

I am also working on independently verifying the hashes here.

tpwrules commented:
I cannot reproduce your hashes. It looks like you built glibc against kernel 6.0 and I built it against kernel 6.1. I made the change which removes the aarch64-linux exception to the default GCC on top of nixpkgs commit `dac3b0d87329edf225ea7d2ed68b9ab79ca8c53b` then built the bootstrap files as described.

It's great that we've shown this is a viable path forward to fix the issue but it's definitely not mergeable as-is. We need reliable provenance for the bootstrap components.

I'm not sure of the protocol for that, looking through the history most times they have been built by Hydra as the "official" versions. But it would be awkward to break half of aarch64-linux for a week trying to get the bootstrap tools rebuilt, even on staging. Someone might be able to create a special branch and job though.

wegank commented:
@tpwrules I'll update the commit once I finished the build on `dac3b0d87329edf225ea7d2ed68b9ab79ca8c53b`. If the hashes still don't match, I guess it might be better to build the bootstrap files on Hydra?

samueldr commented:
> I'm not sure of the protocol for that, looking through the history most times they have been built by Hydra as the "official" versions. But it would be awkward to break half of aarch64-linux for a week trying to get the bootstrap tools rebuilt, even on staging. Someone might be able to create a special branch and job though.

I wonder if this was done through a separate branch being eval'd and built, so it is cached independently from the main channel being affected.

tpwrules commented:
Okay, your new bootstrap files match mine now.

I get a sha256sum of `13eb6afd8357066c766952f32e0c2ec399ba9b350cbce19c97750bcabf2b1bbe` for `bootstrap-tools.tar.xz` and `86e84e9c79a809d24cf57441728349f2632e4e5be3999f57946eb2b7ae695918` for `busybox`.

Please add the nixpkgs hash you used (`dac3b0d87329edf225ea7d2ed68b9ab79ca8c53b`), command (`nix build -f ./pkgs/stdenv/linux/make-bootstrap-tools.nix bootstrapFiles`), and clarification that you made the modification to `pkgs/top-level/all-packages.nix` to the commit message. Looks like that stuff got dropped when you rebased.

I'm not quite ready to declare this the path forward (been researching alternatives today) but if we do, then all that remains I think is to have your files uploaded to tarballs.nixos.org, stick in the official links, and merge this.

amjoseph-nixpkgs commented:
> upgrade to GCC 12 by default which might cause the bootstrap to need to be re-updated

Highly unlikely.  This problem was caused by a major architectural change that ARM made in ARMv8.1.  See [`-moutline-atomics](https://gcc.gnu.org/onlinedocs/gcc/AArch64-Options.html).

The references to `libgcc_s` are a result of providing backward compatibility for ARMv8.0 processors (ie most stuff that isn't in Amazon's data centers).  This was handled similarly to the way soft-float libraries allow backward compatibility for chips without FPUs.

This is a once-a-decade type of change.

tpwrules commented:
Thanks for the additional context. That probably explains why we haven't seen this issue on other architectures.

trofi commented:
> This is a once-a-decade type of change.

I don't think there is any restraint upstream from adding new support helpers there.

`libgcc_s.so.1` has to change every time there are new operations gcc wants to generate out of line. C keeps proliferating with decimal float types, int128 and so on. They all need extra operations that CPU can't do directly. I expect it to happen more in future, not less.

For example even on `x86_64` between `gcc11` and `gcc12` there are quite a few new symbols:

```
$ diff -U0 <(nm -D $(nix-build -A gcc11.cc.lib)/lib/libgcc_s.so.1 | grep -Ev '^ ' | awk '{print $3}' | sort) <(nm -D $(nix-build -A gcc12.cc.lib)/lib/libgcc_s.so.1 | grep -Ev '^ ' | awk '{print $3}' | sort)
--- /dev/fd/63  2023-01-08 13:02:39.743112302 +0000
+++ /dev/fd/62  2023-01-08 13:02:39.744112314 +0000
@@ -25,0 +26 @@
+__divhc3@@GCC_12.0.0
@@ -35,0 +37 @@
+__eqhf2@@GCC_12.0.0
@@ -37,0 +40,5 @@
+__extendhfdf2@@GCC_12.0.0
+__extendhfsf2@@GCC_12.0.0
+__extendhftf2@@GCC_12.0.0
+__extendhfxf2@@GCC_12.0.0
+__extendsfdf2@@GCC_3.0
@@ -42,0 +50 @@
+__fixhfti@@GCC_12.0.0
@@ -48,0 +57 @@
+__fixunshfti@@GCC_12.0.0
@@ -59,0 +69 @@
+__floattihf@@GCC_12.0.0
@@ -65,0 +76 @@
+__floatuntihf@@GCC_12.0.0
@@ -68,0 +80 @@
+GCC_12.0.0
@@ -90,0 +103 @@
+__mulhc3@@GCC_12.0.0
@@ -104,0 +118 @@
+__nehf2@@GCC_12.0.0
@@ -125,0 +140,3 @@
+__truncdfhf2@@GCC_12.0.0
+__truncdfsf2@@GCC_3.0
+__truncsfhf2@@GCC_12.0.0
@@ -126,0 +144 @@
+__trunctfhf2@@GCC_12.0.0
@@ -128,0 +147 @@
+__truncxfhf2@@GCC_12.0.0
```

These are to support `std::bfloat16_t` for example.

amjoseph-nixpkgs commented:
> > This is a once-a-decade type of change.
> 
> I don't think there is any restraint upstream from adding new support helpers there.

Certainly not!  And had no problems with `gcc9`, which is the release that added the The ARM8.1 helpers to `libgcc`.

The problems weren't caused by new symbols appearing in libgcc.  [More details here](https://reviews.llvm.org/D93585#2465024) on gcc10 *defaulting* to `-moutline-atomics`, which did.

> These are to support `std::bfloat16_t` for example.

[`-mfp16-format`] is [off by default](https://gcc.gnu.org/onlinedocs/gcc/Half-Precision.html).

In any event, we should absolutely fix the hideous hackery we're doing with `libgcc_s`.  It's awful, it needs to go:

- https://github.com/NixOS/nixpkgs/pull/209870

trofi commented:
> The problems weren't caused by new symbols appearing in libgcc.

Sure, the problem only happens when `gcc` tries to use them. And new symbols are added specifically to use them by gcc. Usually in the same commit. Either by default or in special modes. The problem is that you can't easily predict if they will be used.

> [-mfp16-format] is [off by default](https://gcc.gnu.org/onlinedocs/gcc/Half-Precision.html).

That is an arm-specific flag. I posted the diff from x86_64 machines. There you might end up using those symbols without extra effort:

```
# on gcc-12
$ printf "int f(_Float16 a, _Float16 b) { return a == b; }" | gcc -S -x c -c - -o - | fgrep hf
        .globl  __extendhfsf2
        call    __extendhfsf2
        call    __extendhfsf2
        call    __extendhfsf2
        call    __extendhfsf2
```

Sure, in this specific case it will be used only by code that mentions `_Float16`. It might sound niche, but both `glibc` (`strtof16()`) and libstdc++ (`<cmath>, `<complex>`) do use it.

Why the final linkage doe not fail on `nixpkgs` if we try to use it in an executable (and whether lack of failure is an intended behaviour) is an exercise for the reader.

wegank commented:
This PR is now superseded by #211007.

wegank commented:
@ofborg build icu

tpwrules commented:
Did you plan to rebuild with GCC 12? It looks at least like Ofborg is happy which is great.

wegank commented:
I think the bootstrap files are sufficient for https://hydra.nixos.org/job/nixpkgs/staging-next/stdenvBootstrapTools.aarch64-linux.dist to pass, so I prefer to have it done by Hydra if it's really necessary.

wegank commented:
I guess I'll have multiple unmaintainable packages after #213324, so I'd like to know if anyone would be happy to upload the tarballs and stick in the official links.

vcunat commented:
**So, is there consensus that we should update the aarch64 bootstrap tarball?**  By that I don't mean to stop improvements to the bootstrapping process itself.  (I'd wait for answers for a few days at least before asking for upload.)

If so, I think I could organize it:
- get Hydra build the tarball with gcc12 (corresponds to a skip of three years)
- redo some basic builds against it (anyone will be able to do that)
- get someone to upload it (I hope we won't get stuck here for too long)
- finally... merge this PR, basically, after refreshing the URL and hash

tpwrules commented:
I think we should update the bootstrap tarball too. We have made great progress on the process updates but there's been some unexpected hitches and software is breaking faster than expected (Rust, Nix itself). It would be good to update, then pull the hacks for those back out. We can then fix the bootstrap with less of a time crunch and debris accumulation.

How exactly do you propose Hydra build the tarball? Can we have it build an arbitrary branch? The problem is if we merge the GCC 9->12 for `aarch64-linux` in order to cause a suitable bootstrap tarball to be built, significant amounts of software will break. It will be fixed again once the new bootstrap tarball is installed but I don't want `aarch64-linux` to be half-broken for a week while that happens.

vcunat commented:
Yes, custom branch is trivial, and here also custom overlay expression to build just the tarball and not all the packages.

wegank commented:
> The problem is if we merge the GCC 9->12 for `aarch64-linux` in order to cause a suitable bootstrap tarball to be built, significant amounts of software will break.

All relevant packages being now on gcc11Stdenv (by doing a treewide search on `/201254`), they should be mostly fine if they already work on x86_64-linux. Of course, someone has to check if they still work with gcc12Stdenv.

I'm also rebuilding the bootstrap tools with GCC 12 on https://github.com/NixOS/nixpkgs/commit/dac3b0d87329edf225ea7d2ed68b9ab79ca8c53b. Let me know if anyone prefers a newer commit.

vcunat commented:
_Maybe_ current `staging-next`.  There's updated binutils where it feels like some of the new features might become useful in future: https://lists.gnu.org/archive/html/info-gnu/2023-01/msg00003.html  (I mean, when we enable stuff in stdenv, it can be cumbersome if bootstrap versions don't support it.)

wegank commented:
> _Maybe_ current `staging-next`. There's updated binutils where it feels like some of the new features might become useful in future: https://lists.gnu.org/archive/html/info-gnu/2023-01/msg00003.html (I mean, when we enable stuff in stdenv, it can be cumbersome if bootstrap versions don't support it.)

Rebuilt against https://github.com/NixOS/nixpkgs/commit/7244b4babe85d158883c3bf949abf7947a050f38.

tpwrules commented:
Was able to build the bootstrap components, verify all the hashes match, and build the original problem children `pkgs.dejagnu` and `pkgs.icu`. Will kick off a full system rebuild overnight on this branch.

@vcunat Can you set us up with the Hydra job to get the bootstrap components built officially?

vcunat commented:
Yes, incidentally I started looking at this a while before you asked.  After a couple attempts I got it right (I think), so it's already building at https://hydra.nixos.org/eval/1790535

vcunat commented:
Done and the two hashes match what's committed here in the PR.

tpwrules commented:
Excellent. I will update with my build results. As I recall @lovesegfault has the privilege to upload the new components?

vcunat commented:
@lovesegfault: so, let me ask explicitly for the upload.

Target paths are a bit trickier than usual.  Customarily we use name of the commit from which the new tools were generated and which was already merged at that point.  Here we need to switch the gcc version first to build the bootstrap tools, but that commit isn't well usable until the new tools get used.  (That's why aarch64-linux has been stuck on old gcc in the first place.)  Still, I think we can include such a commit in history; in particular I'll suggest this commit and upload path: https://github.com/vcunat/nixpkgs/commit/7c026bae62d29600b42359b370b620602e8c498e

tpwrules commented:
Built my VM test environment which is a stock Xfce + Firefox + a few miscellaneous daemons and packages. Everything seems to build and work except for two simple issues:
* `orc` has a [known test failure](https://gitlab.freedesktop.org/gstreamer/orc/-/issues/41) on aarch64 and GCC 12 quite possibly due to a GCC bug. Just disabled the tests and it seems to work fine.
* `vte` won't link under `clangStdenv` anymore due to some C++ string ABI mismatch, reverting that change and just using the regular now-GCC-12 stdenv and it works.

vcunat commented:
Hmm, let me try pinging again: @lovesegfault  (I've tried other people with permissions in the meantime.)

lovesegfault commented:
I don't have access to my AWS keys to upload right now, since I'm traveling, and it's not on my work laptop.

I'm home next week, so that's the earliest I'd be able to help.

I'll try to see if I have them in a backup or something, if so, I can do it today.

Furthermore, I do need a bit more detail to do the upload, just because it's very hard not to mess up. Here's an example of a good request: https://github.com/NixOS/nixpkgs/pull/183487

wegank commented:
> Furthermore, I do need a bit more detail to do the upload, just because it's very hard not to mess up. Here's an example of a good request: #183487

I've put the hashes in the description of changes. This should roughly translate to

```
$ sha256sum ./bootstrap-tools.tar.xz ./busybox
689bedb167ae4076dbd7804667612238ace3427e3a877c7776eb8f11ac06d1e1  ./bootstrap-tools.tar.xz
8ed00fc4b992c3a0a8537fc23d9cd983843c87ad440df6f3e72657255d5c3a6e  ./busybox
```

The targeted commit is `7244b4babe85d158883c3bf949abf7947a050f38`.

Let me know if you need more information.

vcunat commented:
This upcoming week would be nice, given the other (non-)responses I've got so far.

Perhaps you missed some description in the commit which I had linked and proposed to merge later (https://github.com/vcunat/nixpkgs/commit/7c026bae62d29600b42359b370b620602e8c498e).  I prefer to have these important bits directly in there, signed and not (just) in some discussion thread which might be harder to find or even disappear.

You don't need so complicated downloads from Hydra and adding.   I think it's much better to directly get `/nix/store` paths from cache by `nix-store --realize $path` or `nix-build $path`.  (I had thought that obvious and didn't add it into the commit message.)  Anyway, ask for whatever you're missing.

wegank commented:
> Alright, I've uploaded them:

Thank you! I guess these files should be available soon from the links below?

```
http://tarballs.nixos.org/stdenv-linux/aarch64/7244b4babe85d158883c3bf949abf7947a050f38/busybox
http://tarballs.nixos.org/stdenv-linux/aarch64/7244b4babe85d158883c3bf949abf7947a050f38/bootstrap-tools.tar.xz
```

lovesegfault commented:
> > Alright, I've uploaded them:
> 
> Thank you! I guess these files should be available soon from the links below?
> 
> ```
> http://tarballs.nixos.org/stdenv-linux/aarch64/7244b4babe85d158883c3bf949abf7947a050f38/busybox
> http://tarballs.nixos.org/stdenv-linux/aarch64/7244b4babe85d158883c3bf949abf7947a050f38/bootstrap-tools.tar.xz
> ```

Yup! It can take a couple of hours for them to pop-up on the http endpoint, for whatever reason.

lovesegfault commented:
Okay, the issue was: I had forgotten to set the ACL to public-read.

Now fixed, and the links are working. Once you update this PR with the right bootstrap links, I think it can be landed?

K900 commented:
Anyone want to go through the GCC9 workarounds after this?

vcunat commented:
Well, different commit/URL than I posted and explained, but it's just a convention and the files are the same, so I don't really care.

lovesegfault commented:
> Well, different commit/URL than I posted and explained, but it's just a convention and the files are the same, so I don't really care.

Well, that's what I get for trying to do anything with the flu :sneezing_face: 

If you tell me where you want them, I can re-upload to the correct place.

Sorry for the mistake, wanted to get this done ASAP and ended up messing up.

wegank commented:
> If you tell me where you want them, I can re-upload to the correct place.

The commit would be `21ec906463ea8f11abf3f9091ddd4c3276516e58`, with the same files.

lovesegfault commented:
> > If you tell me where you want them, I can re-upload to the correct place.
> 
> The commit would be `21ec906463ea8f11abf3f9091ddd4c3276516e58`, with the same files. Sorry about that!

```
❯ aws s3 ls --recursive s3://nixpkgs-tarballs/stdenv-linux/aarch64/21ec906463ea8f11abf3f9091ddd4c3276516e58/
2023-02-26 15:57:34   20461272 stdenv-linux/aarch64/21ec906463ea8f11abf3f9091ddd4c3276516e58/bootstrap-tools.tar.xz
2023-02-26 15:57:34     257384 stdenv-linux/aarch64/21ec906463ea8f11abf3f9091ddd4c3276516e58/busybox
```

Hopefully, I got it right now, heading to sleep.

wegank committed:
stdenv: aarch64-linux: gcc9 -> gcc12

Hydra job building them: https://hydra.nixos.org/build/208909151

The bootstrap files can be reproduced on the commit 21ec906463ea8f11abf3f9091ddd4c3276516e58, e.g. by:
  cat $(nix-build pkgs/top-level/release.nix -QA stdenvBootstrapTools.aarch64-linux.dist)/nix-support/hydra-build-products
    file tarball /nix/store/kdpbw0plmjqlafjnpbz31ja51m4bd2dk-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz
    file busybox /nix/store/kdpbw0plmjqlafjnpbz31ja51m4bd2dk-stdenv-bootstrap-tools/on-server/busybox
and the hashes as well:
  nix hash file /nix/store/kdpbw0plmjqlafjnpbz31ja51m4bd2dk-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz
    sha256-aJvtsWeuQHbb14BGZ2EiOKzjQn46h3x3duuPEawG0eE=
  nix hash path /nix/store/kdpbw0plmjqlafjnpbz31ja51m4bd2dk-stdenv-bootstrap-tools/on-server/busybox
    sha256-0MuIeQlBUaeisqoFSu8y+8oB6K4ZG5Lhq8RcS9JqkFQ=

You can check this on any machine, as the builds are on cache.nixos.org
but also you can reproduce the hashes when rebuilt on aarch64-linux HW.

wegank commented:
@ofborg build icu

vcunat commented:
Yes, I think so, after another look.
